src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1138 @@
+   if(mEffectComponent->setState(processorState) != kResultOk)
+      return false;
+
+   if(PluginSettings::HasConfigValue(*this, PluginSettings::Private, name, wxT("ProcessorState")))
+   {
+      wxString controllerStateStr;
+      if(!GetConfig(*this, PluginSettings::Private, name, wxT("ControllerState"), controllerStateStr, wxEmptyString))
+         return false;
+      auto controllerState = owned(PresetsBufferStream::fromString(controllerStateStr));
+
+      if(mEditController->setComponentState(processorState) != kResultOk ||
+         mEditController->setState(controllerState) != kResultOk)
+         return false;
+   }
+   SyncParameters();
+
+   return true;
+}
+
+bool VST3Effect::SaveUserPreset(const RegistryPath& name)
+{
+   using namespace Steinberg;
+
+   if(!mEditController)
+      return false;
+
+   auto processorState = owned(safenew PresetsBufferStream);
+   if(mEffectComponent->getState(processorState) != kResultOk)
+      return false;
+
+   SetConfig(*this, PluginSettings::Private, name, wxT("ProcessorState"), processorState->toString());
+
+   auto controllerState = owned(safenew PresetsBufferStream);
+   if(mEditController->getState(controllerState) == kResultOk)
+      SetConfig(*this, PluginSettings::Private, name, wxT("ControllerState"), controllerState->toString());
+
+   return true;
+}
+
+RegistryPaths VST3Effect::GetFactoryPresets()
+{
+   if(!mRescanFactoryPresets)
+      return mFactoryPresets;
+
+   wxArrayString paths;
+   wxDir::GetAllFiles(GetFactoryPresetsPath(mEffectClassInfo), &paths);
+
+   RegistryPaths result;
+   for(auto& path : paths)
+   {
+      wxFileName filename(path);
+      result.push_back(filename.GetName());
+   }
+   mFactoryPresets = std::move(result);
+   mRescanFactoryPresets = false;
+
+   return mFactoryPresets;
+}
+
+bool VST3Effect::LoadFactoryPreset(int id)
+{
+   if(id >= 0 && id < mFactoryPresets.size())
+   {
+      auto filename = wxFileName(GetFactoryPresetsPath(mEffectClassInfo), mFactoryPresets[id] + ".vstpreset");
+      return LoadPreset(filename.GetFullPath());
+   }
+   return true;
+}
+
+bool VST3Effect::LoadFactoryDefaults()
+{
+   using namespace Steinberg;
+   if(mComponentHandler == nullptr)
+      return false;
+
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         if(parameterInfo.flags & Vst::ParameterInfo::kIsReadOnly)
+            continue;
+
+         if(mComponentHandler->beginEdit(parameterInfo.id) == kResultOk)
+         {
+            auto cleanup = finally([&]{
+               mComponentHandler->endEdit(parameterInfo.id);
+            });
+            mComponentHandler->performEdit(parameterInfo.id, parameterInfo.defaultNormalizedValue);
+         }
+         mEditController->setParamNormalized(parameterInfo.id, parameterInfo.defaultNormalizedValue);
```

Comment:
Which is correct, doing this call before or after endEdit?  It changed since my first round of review here, but is consistent with the other call to it now.

Reply:
Doesn't matter actually, because EditController should not notify ComponentHandler about parameter change when seyParamNormalized is invoked directly